crypto/evp/evp_lib.c

```diff
@@ -19,31 +19,39 @@
 #include "evp_locl.h"
 
 #if !defined(FIPS_MODE)
+static int asn1_flags(const EVP_CIPHER *cipher)
+{
+    /*
+     * If this is a provided implementation and EVP_CIPH_FLAG_CUSTOM_ASN1 is
+     * set, the implementation has its own handling of AlgorithmIdentifier
+     * parameters.
+     * If EVP_CIPH_FLAG_DEFAULT_ASN1 is set, the implementation is asking
+     * libcrypto to handle the ASN.1 params with default libcrypto methods.
+     * Otherwise.  This is a legacy flag.
+     */
+    int flags = EVP_CIPHER_flags(cipher);
+
+    /*
+     * If EVP_CIPH_FLAG_DEFAULT_ASN1 is unset, we set
+     * EVP_CIPH_FLAG_CUSTOM_ASN1.
+     * If EVP_CIPH_FLAG_DEFAULT_ASN1 is set, we ignore it, giving
+     * EVP_CIPH_FLAG_CUSTOM_ASN1 higher priority.
+     */
+    if ((flags & EVP_CIPH_FLAG_DEFAULT_ASN1) == 0)
```

Comment:
Doesn't there need to be a check here to see if the cipher is provided or not? Provided ciphers won't have `EVP_CIPH_FLAG_DEFAULT_ASN1` and don't necessarily want `EVP_CIPH_FLAG_CUSTOM_ASN1`

Reply:
Hmmmm...  I dunno, do we?  What we *actually* need to do is figure out what assumption we should go for with the following conditions:

1.  neither `EVP_CIPH_FLAG_DEFAULT_ASN1` nor `EVP_CIPH_FLAG_CUSTOM_ASN1` are set, i.e. what default do we assume?  Does that depend on the type of implementation?
2.  both `EVP_CIPH_FLAG_DEFAULT_ASN1` and `EVP_CIPH_FLAG_CUSTOM_ASN1` are set, i.e. which flag has higher priority?  Does that depend on the type of implementation?

I'm actually soooo sooooo sooooo tempted to make `EVP_CIPH_FLAG_DEFAULT_ASN1` just be zero and only check for `EVP_CIPH_FLAG_CUSTOM_ASN1` regardless.  That would take away the problem...  I know that some legacy implementations aren't setting this flag even though they should, *but* they're calling the exact same functions that would have been called by `EVP_CIPHER_{param_to_asn1,asn1_to_param}`, so they work *in practice* as if that flag had been set.

You know what?  I'll go with that temptation.  It would be source compatible.

Reply:
As I wrote in the initial comment my suggested behaviour was:

> In other words, you get default ASN1 behaviour if:
> 
> You have a legacy cipher and EVP_CIPH_FLAG_DEFAULT_ASN1 is set
> You have a provided cipher and EVP_CIPH_FLAG_NON_DEFAULT_ASN1 is not set

So, yes, it does depend on the type of implementation. Of course that is temporary for our own ciphers since we're moving everything to be a provided cipher anyway. We'll still have to handle legacy uses via EVP_CIPHER_meth_new() et al though.

Reply:
I realised that there's no need to check `EVP_CIPH_FLAG_DEFAULT_ASN1` at all.
legacy implementations that don't use this flag MUST set `set_asn1_parameters` / `get_asn1_parameters`, so it's enough to check that these pointers are non-NULL.

If those pointers are NULL, it's sufficient to check for `EVP_CIPH_FLAG_CUSTOM_ASN1`, and *if it is set*, also check that `cipher->prov != NULL`, and act appropriately in those two cases.

So I ended up reimplementing these functions again.  Local test going, will show up soon.